recipes/generate.py

```diff
@@ -0,0 +1,176 @@
+            logits = torch.where(logits < pivot, -float("Inf"), logits)
+        probs = torch.nn.functional.softmax(logits, dim=-1)
+        return probs
+
+    def _sample(self, logits: torch.Tensor, temperature: float, top_k: int):
+        probs = self._logits_to_probs(logits[0, -1], temperature, top_k)
+        idx_next = self._multinomial_sample_one_no_sync(probs)
+        return idx_next, probs
+
+    def _prefill(
+        self,
+        model: nn.Module,
+        x: torch.Tensor,
+        input_pos: torch.Tensor,
+        temperature: float,
+        top_k: int,
+    ) -> torch.Tensor:
+        # input_pos: [B, S]
+        logits = model(x, input_pos)
+        return self._sample(logits, temperature, top_k)[0]
+
+    def _decode_one_token(
+        self,
+        model: nn.Module,
+        x: torch.Tensor,
+        input_pos: torch.Tensor,
+        temperature: float,
+        top_k: int,
+    ):
+        # input_pos: [B, 1]
+        assert input_pos.shape[-1] == 1
+        logits = model(x, input_pos)
+        return self._sample(logits, temperature, top_k)
+
+    def _decode_n_tokens(
+        self,
+        model: nn.Module,
+        cur_token: torch.Tensor,
+        input_pos: torch.Tensor,
+        num_new_tokens: int,
+        temperature: float,
+        top_k: int,
+    ):
+        new_tokens, new_probs = [], []
+        for i in range(num_new_tokens):
+            with torch.backends.cuda.sdp_kernel(
+                enable_flash=False, enable_mem_efficient=False, enable_math=True
+            ):
+                next_token, next_prob = self._decode_one_token(
+                    model, cur_token, input_pos, temperature, top_k
+                )
+                input_pos += 1
+                new_tokens.append(next_token.clone())
+                new_probs.append(next_prob.clone())
+                cur_token = next_token.view(1, -1)
+        return new_tokens, new_probs
+
+    @torch.no_grad()
+    def generate(self, cfg: DictConfig):
+        tokens = self._tokenizer.encode(cfg.prompt, add_bos=True, add_eos=False)
+        prompt = torch.tensor(tokens, dtype=torch.int, device=self._device)
+
+        prompt_length = prompt.size(0)
+        total_length = prompt_length + cfg.max_new_tokens
+
+        empty = torch.empty(total_length, dtype=torch.int, device=self._device)
+        empty[:prompt_length] = prompt
+        seq = empty
+        input_pos = torch.arange(0, prompt_length, device=self._device)
+
+        next_token = self._prefill(
+            self._model,
+            prompt.view(1, -1),
+            input_pos,
+            temperature=cfg.temperature,
+            top_k=cfg.top_k,
+        )
+        seq[prompt_length] = next_token
+        input_pos = torch.tensor([prompt_length], device=self._device, dtype=torch.int)
+
+        t0 = time.perf_counter()
+        generated_tokens, _ = self._decode_n_tokens(
+            self._model,
+            next_token.view(1, -1),
```

Comment:
What is the view actually doing here? Isn't this just a single token?

Reply:
Follow-up.. just a general comment on a bunch of these methods: it'd be nice to add docstrings with examples of each one (e.g. given a prompt of length n, prefill will fill the KV cache's first n elements and return token n+1, something similar for `_decode_one_token` and/or `_decode_n_tokens`)